src/main/kotlin/org/rust/bsp/toolwindow/actions/BuildAction.kt

```diff
@@ -0,0 +1,58 @@
+package org.rust.bsp.toolwindow.actions
+
+import ch.epfl.scala.bsp4j.BuildTargetIdentifier
+import ch.epfl.scala.bsp4j.CompileParams
+import com.intellij.icons.AllIcons
+import com.intellij.openapi.actionSystem.ActionUpdateThread
+import com.intellij.openapi.actionSystem.AnAction
+import com.intellij.openapi.actionSystem.AnActionEvent
+import com.intellij.openapi.components.service
+import com.intellij.openapi.diagnostic.logger
+import com.intellij.openapi.progress.runBackgroundableTask
+import com.intellij.openapi.progress.runModalTask
+import com.intellij.openapi.project.Project
+import org.rust.bsp.service.BspConnectionService
+
+class BuildAction(
+    val target: String
+) : AnAction("Build Target", null, AllIcons.Actions.Compile) {
+
+    override fun actionPerformed(e: AnActionEvent) {
+        val project = e.project
+
+        if (project != null) {
+            doAction(project)
+        } else {
+            log.warn("BuildAction cannot be performed! Project not available.")
+        }
+    }
+
+    private fun doAction(project: Project) {
+        runBackgroundableTask("Building...", project = project, cancellable = false) {
+            val connection = project.service<BspConnectionService>()
+            connection.compileSolution(CompileParams(listOf(BuildTargetIdentifier(target))))
```

Comment:
Can `connection` be null?

Reply:
According to documentation project.service() always returns an instance, or creates new one - because of that it is impossible for it to be null